File: repo-templates/go-lambda-webapp/skeleton/internal/orders/order_test.go
package orders

import "testing"

func TestValidate(t *testing.T) {
	good := []Item{{SKU: "SKU-1", Qty: 2, Price: 9.99}}

	cases := map[string]struct {
		order Order
		want  string
	}{
		"a complete order is accepted": {
			order: Order{Customer: "ada", Items: good},
		},
		"an order with no customer is rejected": {
			order: Order{Items: good},
			want:  "customer is required",
		},
		"an order with no items is rejected": {
			order: Order{Customer: "ada"},
			want:  "an order needs at least one item",
		},
		"an item with no sku is rejected": {
			order: Order{Customer: "ada", Items: []Item{{Qty: 1, Price: 1}}},
			want:  "items[0]: sku is required",
		},
		"an item with zero quantity is rejected": {
			order: Order{Customer: "ada", Items: []Item{{SKU: "SKU-1", Qty: 0, Price: 1}}},
			want:  "items[0]: qty must be at least 1, got 0",
		},
		"an item with a negative price is rejected": {
			order: Order{Customer: "ada", Items: []Item{{SKU: "SKU-1", Qty: 1, Price: -1}}},
			want:  "items[0]: price cannot be negative, got -1",
		},
	}

	for name, tc := range cases {
		t.Run(name, func(t *testing.T) {
			err := tc.order.Validate()
			switch {
			case tc.want == "" && err != nil:
				t.Fatalf("want no error, got %v", err)
			case tc.want != "" && err == nil:
				t.Fatalf("want error %q, got none", tc.want)
			case tc.want != "" && err.Error() != tc.want:
				t.Fatalf("want error %q, got %q", tc.want, err)
			}
		})
	}
}

// Totalise is what stops a client stating its own total, so a submitted Total
// has to be overwritten rather than merely filled in when absent.
func TestTotaliseOverwritesASubmittedTotal(t *testing.T) {
	o := Order{
		Customer: "ada",
		Items:    []Item{{SKU: "SKU-1", Qty: 3, Price: 2.50}, {SKU: "SKU-2", Qty: 1, Price: 0.50}},
		Total:    0.01,
	}
	o.Totalise()
	if o.Total != 8.00 {
		t.Fatalf("want total 8.00, got %v", o.Total)
	}
}

File: repo-templates/java-lambda-webapp/skeleton/src/test/java/com/example/orders/OrderTest.java
package com.example.orders;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.time.Instant;
import java.util.List;

import org.junit.jupiter.api.Test;

class OrderTest {

    private static Order order(String customer, List<Item> items) {
        return new Order(null, customer, items, 0, null);
    }

    @Test
    void validOrderHasNoProblem() {
        assertNull(order("ada", List.of(new Item("SKU-1", 1, 0))).problem());
    }

    @Test
    void reportsTheFirstProblemInTheCallersWords() {
        assertEquals("customer is required", order("", List.of(new Item("SKU-1", 1, 1))).problem());
        assertEquals("an order needs at least one item", order("ada", List.of()).problem());
        assertEquals("items[0]: sku is required", order("ada", List.of(new Item("", 1, 1))).problem());
        assertEquals("items[1]: qty must be at least 1, got 0",
                order("ada", List.of(new Item("A", 1, 1), new Item("B", 0, 1))).problem());
        assertEquals("items[0]: price cannot be negative, got -1.0",
                order("ada", List.of(new Item("A", 1, -1))).problem());
    }

    // A client that states its own total must not have it believed.
    @Test
    void acceptedOverwritesASubmittedTotal() {
        Order submitted = new Order("theirs", "ada", List.of(new Item("A", 2, 1.5), new Item("B", 1, 4)), 999, null);
        Instant at = Instant.parse("2026-01-02T03:04:05Z");
        Order saved = submitted.accepted("ours", at);
        assertEquals(7.0, saved.total());
        assertEquals("ours", saved.orderId());
        assertEquals(at, saved.receivedAt());
    }
}
